plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/inspector-controls/hand-picked-products-control.tsx

```diff
@@ -36,13 +42,27 @@ function useProducts() {
 	);
 
 	useEffect( () => {
-		getProducts( {
-			selected: [],
-			queryArgs: {
-				// Fetch all products.
-				per_page: 0,
-			},
-		} ).then( ( results ) => {
+		// We take two strategies here because of internal logic of
+		// `getProducts` and `getProductsRequests` that skips request for
+		// `selected` products for small stores. So fetching products per user's input
+		// breaks selected items:
+		// 1. For large stores (>100 products) we fetch products as input changes AND
+		// `selected` products.
+		// 2. For small stores (<=100 products) we fetch all products just once.
+
+		const query = {
+			selected: isLargeCatalog ? selected.map( Number ) : [],
+			queryArgs: isLargeCatalog
+				? {
+						search,
+						// Limit search to 40 results. If results are not satisfying
+						// user needs to type more characters to get closer to actual
+						// product name.
+						per_page: 40,
+				  }
+				: {},
+		};
+		getProducts( query ).then( ( results ) => {
 			const newProductsMap = new Map();
 			( results as ProductResponseItem[] ).forEach( ( product ) => {
```

Comment:
Again, this one is also not under the scope of this PR, but it's possible to select the wrong product if we have two or more products with the same title:

```
newProductsMap.set( product.name, product );
```